ukbb_qc/annotations/fix_frequency_data.py

```diff
@@ -0,0 +1,197 @@
+"""
+
+TRANCHE_DATA = ("broad", CURRENT_FREEZE)
+"""
+UKB tranche data (data source and data freeze number).
+"""
+
+VDS_PATH = "gs://ukbb-pharma-exome-analysis/500k_temp/500k.vds"
+"""
+Path to Variant Dataset (VDS) that contains high quality samples from the final UK Biobank callset.
+
+Generated using the following commands:
+# NOTE: If ever need to rerun, should read v4 VDS using `get_gnomad_v4_vds`
+# (in gnomad_qc/v4/resoures/basics.py)
+vds = hl.vds.read_vds("gs://gnomad/raw/exomes/4.0/gnomad_v4.0.vds")
+meta_ht = hl.read_table("gs://broad-ukbb/broad.freeze_7/sample_qc/meta.ht")
+meta_ht = meta_ht.filter(meta_ht.sample_filters.high_quality)
+vds = hl.vds.filter_samples(vds, meta_ht, remove_dead_alleles=True)
+call_stats_ht = hl.read_table(ukb.var_annotations_ht_path('ukb_freq', *TRANCHE_DATA[CURRENT_TRANCHE]))
+freq_index = get_cohort_index(call_stats_ht)
+var = vds.variant_data.annotate_rows(call_stats=call_stats_ht[vds.variant_data.row_key].freq[freq_index])
+vds = hl.vds.VariantDataset(vds.reference_data, var)
+vds.write(ukb_exomes_path, overwrite=args.overwrite)
+
+Full VDS size: 8.69 TiB
+VDS variant data size: 561.97 GiB
+"""
+
+
+def generate_frequency_data(
+    mt: hl.MatrixTable,
+) -> hl.Table:
+    """
+    Generates frequency struct annotation containing AC, AF, AN, and homozygote count for input dataset stratified by population.
+
+    Option to stratify by platform (either inferred platform or tranche as proxy for platform).
+
+    .. note::
+
+        This function expects dense format data.
+        Frequency data is generated using gnomAD PC project population labels as the main population stratification.
+        Hybrid population labels are used as the subpopulation stratification.
+        However, cohort frequency (frequency including all related samples) data is generated using hybrid population labels.
+
+    :param MatrixTable mt: Input MatrixTable.
+    :param int freeze: One of the data freezes.
+    :param list pops_to_remove_for_popmax: List of populations to exclude from popmax calculations.
+    :param list cohort_frequency_pops: List of populations to include in cohort frequency calculations.
+    :param bool platform_strata: Whether to calculate frequencies per platform or tranche as proxy for platform.
+    :param bool tranche: Whether to use tranche (as a proxy for platform) instead of inferred platform.
+    :param bool calculate_by_tranche: Whether to calculate frequencies per tranche.
+    :param hl.expr.StringExpression platform_expr: Expression containing platform or tranche information. Required if platform_strata is set.
+    :param str data_source: One of 'regeneron' or 'broad'. Default is 'broad'.
+    :return: MatrixTable with frequency annotations in struct named `freq` and metadata in globals named `freq_meta`.
+    :rtype: hl.MatrixTable
+    """
+    logger.info("Generating frequency data...")
+    mt = annotate_freq(
+        mt,
+        sex_expr=mt.sex_karyotype,
```

Comment:
Can we forgo the sex annotations?

Reply:
oh I'm happy to drop these if you don't need them. should I just run `hl.agg.call_stats` instead then?

Reply:
I just realized I didn't add the adj annotation to the VDS, so I'll add that and keep `annotate_freq` for now (unless you only need `adj` frequencies, in which case maybe we should use `call_stats`)